File: backend/src/models/Device.js
import mongoose from 'mongoose';

const deviceSchema = new mongoose.Schema(
  {
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true,
      index: true,
    },
    deviceId: {
      type: String,
      required: [true, 'Device ID is required'],
      trim: true,
    },
    deviceName: {
      type: String,
      required: [true, 'Device Name is required'],
      trim: true,
      default: 'Android Device',
    },
    platform: {
      type: String,
      default: 'Android',
    },
    lastSeen: {
      type: Date,
      default: Date.now,
    },
  },
  {
    timestamps: true,
  }
);

// Compound unique index ensuring a user cannot have duplicate deviceId registrations
deviceSchema.index({ userId: 1, deviceId: 1 }, { unique: true });

export const Device = mongoose.model('Device', deviceSchema);

File: mobile/lib/models/device.dart
class Device {
  final String deviceId;
  final String deviceName;
  final String platform;
  final DateTime? lastSeen;

  Device({
    required this.deviceId,
    required this.deviceName,
    this.platform = 'Android',
    this.lastSeen,
  });

  factory Device.fromMap(Map<String, dynamic> map) {
    return Device(
      deviceId: map['deviceId']?.toString() ?? '',
      deviceName: map['deviceName']?.toString() ?? 'Android Device',
      platform: map['platform']?.toString() ?? 'Android',
      lastSeen: map['lastSeen'] != null
          ? DateTime.tryParse(map['lastSeen'].toString())
          : null,
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'deviceId': deviceId,
      'deviceName': deviceName,
      'platform': platform,
      if (lastSeen != null) 'lastSeen': lastSeen!.toIso8601String(),
    };
  }
}
